File: app/static/js/nginx_rules.js
/**
 * Nginx Rules Management for Edge Nodes
 * Handles loading, editing, and applying Nginx configuration rules
 */

// ============================================
// Nginx Rules Data Management
// ============================================

// Track loading state
let nginxRulesLoaded = false;

/**
 * Show loading overlay on nginx tab
 */
function showNginxLoading() {
    const nginxTab = document.getElementById('tab-nginx');
    if (!nginxTab) return;
    
    // Create loading overlay if it doesn't exist
    let overlay = document.getElementById('nginx-loading-overlay');
    if (!overlay) {
        overlay = document.createElement('div');
        overlay.id = 'nginx-loading-overlay';
        overlay.className = 'nginx-loading-overlay';
        overlay.innerHTML = `
            <div class="nginx-loading-spinner">
                <i class="fas fa-spinner fa-spin"></i>
                <span>Загрузка настроек с ноды...</span>
            </div>
        `;
        nginxTab.style.position = 'relative';
        nginxTab.appendChild(overlay);
    }
    
    overlay.classList.add('visible');
}

/**
 * Hide loading overlay on nginx tab
 */
function hideNginxLoading() {
    const overlay = document.getElementById('nginx-loading-overlay');
    if (overlay) {
        overlay.classList.remove('visible');
    }
}

/**
 * Load current Nginx rules from the server
 */
async function loadNginxRules() {
    // Show loading overlay
    showNginxLoading();
    
    try {
        const response = await api.get(`/edge-nodes/${NODE_ID}/nginx-rules`);
        if (response && response.config) {
            populateNginxForm(response.config);
            nginxRulesLoaded = true;
        }
    } catch (error) {
        console.error('Failed to load nginx rules:', error);
        // Use defaults if can't load
    } finally {
        // Hide loading overlay after a small delay for smooth transition
        setTimeout(() => {
            hideNginxLoading();
        }, 300);
    }
}

/**
 * Populate form fields with config data
 */
function populateNginxForm(config) {
    // Client settings
    if (config.client) {
        setVal('nginx-client-max-body-size', config.client.client_max_body_size, '100m');
        setVal('nginx-client-body-timeout', config.client.client_body_timeout, 60);
        setVal('nginx-client-header-timeout', config.client.client_header_timeout, 60);
        setVal('nginx-client-body-buffer-size', config.client.client_body_buffer_size, '128k');
    }
    
    // WebSocket settings
    if (config.websocket) {
        setChecked('nginx-ws-enabled', config.websocket.enabled !== false);
        setVal('nginx-ws-read-timeout', config.websocket.read_timeout, 3600);
        setVal('nginx-ws-send-timeout', config.websocket.send_timeout, 3600);
        setVal('nginx-ws-connect-timeout', config.websocket.connect_timeout, 60);
    }
    
    // Proxy settings
    if (config.proxy) {
        setVal('nginx-proxy-connect-timeout', config.proxy.proxy_connect_timeout, 60);
        setVal('nginx-proxy-send-timeout', config.proxy.proxy_send_timeout, 60);
        setVal('nginx-proxy-read-timeout', config.proxy.proxy_read_timeout, 60);
        setVal('nginx-proxy-buffer-size', config.proxy.proxy_buffer_size, '4k');
        setVal('nginx-proxy-buffers', config.proxy.proxy_buffers, '8 4k');
    }
    
    // Gzip settings
    if (config.gzip) {
        setChecked('nginx-gzip-enabled', config.gzip.enabled !== false);
        setVal('nginx-gzip-level', config.gzip.comp_level, 6);
        const levelDisplay = document.getElementById('gzip-level-value');
        if (levelDisplay) levelDisplay.textContent = config.gzip.comp_level || 6;
        setVal('nginx-gzip-min-length', config.gzip.min_length, 1000);
    }
    
    // Keepalive settings
    if (config.keepalive) {
        setVal('nginx-keepalive-timeout', config.keepalive.timeout, 65);
        setVal('nginx-keepalive-requests', config.keepalive.requests, 1000);
        setVal('nginx-keepalive-upstream', config.keepalive.upstream_connections, 32);
    }
    
    // HTTP/2 settings
    if (config.http2) {
        setChecked('nginx-http2-enabled', config.http2.enabled !== false);
        setVal('nginx-http2-streams', config.http2.max_concurrent_streams, 128);
    }
    
    // Rate limiting
    if (config.rate_limit) {
        setChecked('nginx-rate-enabled', config.rate_limit.enabled === true);
        setVal('nginx-rate-limit', config.rate_limit.rate, '100r/s');
        setVal('nginx-rate-burst', config.rate_limit.burst, 200);
    }
    
    // Security settings
    if (config.security) {
        setChecked('nginx-hide-version', config.security.server_tokens === false);
        setChecked('nginx-x-frame-options', config.security.add_x_frame_options !== false);
        setChecked('nginx-x-content-type', config.security.add_x_content_type_options !== false);
        setChecked('nginx-x-xss', config.security.add_x_xss_protection !== false);
    }
    
    // SSL settings
    if (config.ssl) {
        const protocols = config.ssl.protocols || ['TLSv1.2', 'TLSv1.3'];
        setChecked('nginx-tls12', protocols.includes('TLSv1.2'));
        setChecked('nginx-tls13', protocols.includes('TLSv1.3'));
        setChecked('nginx-ocsp-stapling', config.ssl.stapling !== false);
        setVal('nginx-ssl-session-timeout', config.ssl.session_timeout, '1d');
    }
    
    // Cache settings
    if (config.cache) {
        setChecked('nginx-cache-enabled', config.cache.enabled !== false);
        setVal('nginx-cache-max-size', config.cache.max_size, '10g');
        setVal('nginx-cache-inactive', config.cache.inactive, '7d');
    }
}

// Helper functions for form population
function setVal(id, value, defaultValue) {
    const el = document.getElementById(id);
    if (el) el.value = value !== undefined && value !== null ? value : defaultValue;
}

function setChecked(id, checked) {
    const el = document.getElementById(id);
    if (el) el.checked = checked;
}

function getVal(id) {
    const el = document.getElementById(id);
    return el ? el.value : '';
}

function getInt(id) {
    return parseInt(getVal(id)) || 0;
}

function isChecked(id) {
    const el = document.getElementById(id);
    return el ? el.checked : false;
}

/**
 * Collect all form values into Nginx rules config object
 */
function collectNginxRules() {
    const protocols = [];
    if (isChecked('nginx-tls12')) protocols.push('TLSv1.2');
    if (isChecked('nginx-tls13')) protocols.push('TLSv1.3');
    
    return {
        client: {
            client_max_body_size: getVal('nginx-client-max-body-size'),
            client_body_timeout: getInt('nginx-client-body-timeout'),
            client_header_timeout: getInt('nginx-client-header-timeout'),
            client_body_buffer_size: getVal('nginx-client-body-buffer-size'),
            large_client_header_buffers: "4 16k"
        },
        websocket: {
            enabled: isChecked('nginx-ws-enabled'),
            read_timeout: getInt('nginx-ws-read-timeout'),
            send_timeout: getInt('nginx-ws-send-timeout'),
            connect_timeout: getInt('nginx-ws-connect-timeout')
        },
        proxy: {
            proxy_connect_timeout: getInt('nginx-proxy-connect-timeout'),
            proxy_send_timeout: getInt('nginx-proxy-send-timeout'),
            proxy_read_timeout: getInt('nginx-proxy-read-timeout'),
            proxy_buffer_size: getVal('nginx-proxy-buffer-size'),
            proxy_buffers: getVal('nginx-proxy-buffers'),
            proxy_busy_buffers_size: "8k"
        },
        gzip: {
            enabled: isChecked('nginx-gzip-enabled'),
            comp_level: getInt('nginx-gzip-level'),
            min_length: getInt('nginx-gzip-min-length'),
            types: [
                "text/plain", "text/css", "text/javascript",
                "application/javascript", "application/json",
                "application/xml", "image/svg+xml"
            ],
            vary: true
        },
        ssl: {
            protocols: protocols,
            prefer_server_ciphers: true,
            session_timeout: getVal('nginx-ssl-session-timeout'),
            session_cache: "shared:SSL:50m",
            stapling: isChecked('nginx-ocsp-stapling')
        },
        rate_limit: {
            enabled: isChecked('nginx-rate-enabled'),
            zone_name: "cdn_limit",
            zone_size: "10m",
            rate: getVal('nginx-rate-limit'),
            burst: getInt('nginx-rate-burst'),
            nodelay: true
        },
        cache: {
            enabled: isChecked('nginx-cache-enabled'),
            path: "/var/cache/nginx/cdn",
            zone_name: "cdn_cache",
            zone_size: "100m",
            max_size: getVal('nginx-cache-max-size'),
            inactive: getVal('nginx-cache-inactive'),
            use_stale: ["error", "timeout", "updating", "http_500", "http_502", "http_503", "http_504"],
            valid_codes: {"200": "1d", "301": "1h", "302": "1h", "404": "1m"}
        },
        keepalive: {
            timeout: getInt('nginx-keepalive-timeout'),
            requests: getInt('nginx-keepalive-requests'),
            upstream_connections: getInt('nginx-keepalive-upstream')
        },
        http2: {
            enabled: isChecked('nginx-http2-enabled'),
            max_concurrent_streams: getInt('nginx-http2-streams'),
            max_field_size: "4k",
            max_header_size: "16k"
        },
        security: {
            server_tokens: !isChecked('nginx-hide-version'),
            add_x_frame_options: isChecked('nginx-x-frame-options'),
            add_x_content_type_options: isChecked('nginx-x-content-type'),
            add_x_xss_protection: isChecked('nginx-x-xss')
        }
    };
}

// ============================================
// Nginx Rules Actions
// ============================================

/**
 * Apply Nginx rules to the edge node
 */
async function applyNginxRules() {
    const rules = collectNginxRules();
    
    showNotification('Применение правил Nginx...', 'info');
    
    try {
        const response = await api.put(`/edge-nodes/${NODE_ID}/nginx-rules`, rules);
        
        if (response.success) {
            showNotification('Правила Nginx применены успешно!', 'success');
            if (typeof log === 'function') {
                log('Nginx rules applied successfully', 'success');
                if (response.config_test_output) {
                    log(response.config_test_output);
                }
            }
        } else {
            showNotification(`Ошибка: ${response.message}`, 'error');
            if (typeof log === 'function') {
                log(`Failed to apply nginx rules: ${response.message}`, 'error');
                if (response.config_test_output) {
                    log(response.config_test_output, 'error');
                }
            }
        }
    } catch (error) {
        showNotification(`Ошибка: ${error.message}`, 'error');
        if (typeof log === 'function') {
            log(`Error applying nginx rules: ${error.message}`, 'error');
        }
    }
}

/**
 * Test Nginx configuration without applying
 */
async function testNginxRules() {
    const rules = collectNginxRules();
    
    showNotification('Тестирование конфигурации...', 'info');
    
    try {
        const response = await api.put(`/edge-nodes/${NODE_ID}/nginx-rules?test_only=true`, rules);
        
        if (response.success) {
            showNotification('Конфигурация корректна!', 'success');
            if (typeof log === 'function' && response.config_test_output) {
                log('Nginx config test: ' + response.config_test_output, 'success');
            }
        } else {
            showNotification(`Ошибка конфигурации: ${response.message}`, 'error');
            if (typeof log === 'function' && response.config_test_output) {
                log('Config test failed: ' + response.config_test_output, 'error');
            }
        }
    } catch (error) {
        showNotification(`Ошибка: ${error.message}`, 'error');
    }
}

/**
 * Reset Nginx rules to defaults
 */
async function resetNginxRules() {
    if (!confirm('Сбросить все настройки Nginx на значения по умолчанию?')) {
        return;
    }
    
    showNotification('Сброс настроек...', 'info');
    
    try {
        const response = await api.post(`/edge-nodes/${NODE_ID}/nginx-rules/reset`);
        
        if (response.success) {
            showNotification('Настройки сброшены', 'success');
            await loadNginxRules();
        } else {
            showNotification(`Ошибка: ${response.message}`, 'error');
        }
    } catch (error) {
        showNotification(`Ошибка: ${error.message}`, 'error');
    }
}

/**
 * Preview generated Nginx configuration
 */
async function previewNginxConfig() {
    const rules = collectNginxRules();
    const previewEl = document.getElementById('nginx-config-preview');
    
    if (!previewEl) return;
    
    try {
        const response = await api.post(`/edge-nodes/${NODE_ID}/nginx-rules/preview`, rules);
        
        if (response.main_config) {
            previewEl.textContent = response.main_config + 
                '\n\n# === Location snippet (include in server blocks) ===\n' + 
                response.location_snippet;
            previewEl.style.display = 'block';
        }
    } catch (error) {
        previewEl.textContent = `Error: ${error.message}`;
        previewEl.style.display = 'block';
    }
}

// ============================================
// UI Initialization
// ============================================

/**
 * Initialize Nginx rules UI elements
 */
function initNginxRulesUI() {
    // Update gzip level display on slider change
    const gzipSlider = document.getElementById('nginx-gzip-level');
    if (gzipSlider) {
        gzipSlider.addEventListener('input', function() {
            const display = document.getElementById('gzip-level-value');
            if (display) display.textContent = this.value;
        });
    }
}

// Initialize when DOM is ready
if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', initNginxRulesUI);
} else {
    initNginxRulesUI();
}

File: app/schemas/nginx_rules.py
"""Nginx rules schemas for edge node configuration"""
from typing import Optional, Dict, Any, List
from pydantic import BaseModel, Field


class NginxClientSettings(BaseModel):
    """Client connection settings"""
    client_max_body_size: str = Field(
        default="100m",
        description="Maximum allowed size of the client request body (e.g., 100m, 1g)"
    )
    client_body_timeout: int = Field(
        default=60,
        ge=1, le=3600,
        description="Timeout for reading client request body (seconds)"
    )
    client_header_timeout: int = Field(
        default=60,
        ge=1, le=3600,
        description="Timeout for reading client request header (seconds)"
    )
    client_body_buffer_size: str = Field(
        default="128k",
        description="Buffer size for reading client request body"
    )
    large_client_header_buffers: str = Field(
        default="4 16k",
        description="Number and size of buffers for large client headers"
    )


class NginxWebSocketSettings(BaseModel):
    """WebSocket proxy settings"""
    enabled: bool = Field(
        default=True,
        description="Enable WebSocket protocol support"
    )
    read_timeout: int = Field(
        default=3600,
        ge=1, le=86400,
        description="Proxy read timeout for WebSocket connections (seconds)"
    )
    send_timeout: int = Field(
        default=3600,
        ge=1, le=86400,
        description="Proxy send timeout for WebSocket connections (seconds)"
    )
    connect_timeout: int = Field(
        default=60,
        ge=1, le=300,
        description="Timeout for establishing connection to upstream"
    )


class NginxProxySettings(BaseModel):
    """Proxy settings for backend connections"""
    proxy_connect_timeout: int = Field(
        default=60,
        ge=1, le=300,
        description="Timeout for establishing connection to backend (seconds)"
    )
    proxy_send_timeout: int = Field(
        default=60,
        ge=1, le=3600,
        description="Timeout for transmitting request to backend (seconds)"
    )
    proxy_read_timeout: int = Field(
        default=60,
        ge=1, le=3600,
        description="Timeout for reading response from backend (seconds)"
    )
    proxy_buffer_size: str = Field(
        default="4k",
        description="Size of the buffer for reading first part of response"
    )
    proxy_buffers: str = Field(
        default="8 4k",
        description="Number and size of buffers for reading response"
    )
    proxy_busy_buffers_size: str = Field(
        default="8k",
        description="Maximum size of busy buffers"
    )


class NginxGzipSettings(BaseModel):
    """Gzip compression settings"""
    enabled: bool = Field(
        default=True,
        description="Enable gzip compression"
    )
    comp_level: int = Field(
        default=6,
        ge=1, le=9,
        description="Compression level (1-9, higher = more compression)"
    )
    min_length: int = Field(
        default=1000,
        ge=0,
        description="Minimum response length to compress (bytes)"
    )
    types: List[str] = Field(
        default=[
            "text/plain",
            "text/css",
            "text/javascript",
            "application/javascript",
            "application/json",
            "application/xml",
            "image/svg+xml"
        ],
        description="MIME types to compress"
    )
    vary: bool = Field(
        default=True,
        description="Add Vary: Accept-Encoding header"
    )


class NginxSSLSettings(BaseModel):
    """SSL/TLS settings"""
    protocols: List[str] = Field(
        default=["TLSv1.2", "TLSv1.3"],
        description="Enabled TLS protocols"
    )
    prefer_server_ciphers: bool = Field(
        default=True,
        description="Prefer server ciphers over client ciphers"
    )
    session_timeout: str = Field(
        default="1d",
        description="SSL session cache timeout"
    )
    session_cache: str = Field(
        default="shared:SSL:50m",
        description="SSL session cache configuration"
    )
    stapling: bool = Field(
        default=True,
        description="Enable OCSP stapling"
    )


class NginxRateLimitSettings(BaseModel):
    """Rate limiting settings"""
    enabled: bool = Field(
        default=False,
        description="Enable rate limiting"
    )
    zone_name: str = Field(
        default="cdn_limit",
        description="Name of the rate limit zone"
    )
    zone_size: str = Field(
        default="10m",
        description="Size of the rate limit zone"
    )
    rate: str = Field(
        default="100r/s",
        description="Request rate limit (e.g., 100r/s, 10r/m)"
    )
    burst: int = Field(
        default=200,
        ge=0,
        description="Maximum burst of requests"
    )
    nodelay: bool = Field(
        default=True,
        description="Process burst requests immediately"
    )


class NginxCacheSettings(BaseModel):
    """Proxy cache settings"""
    enabled: bool = Field(
        default=True,
        description="Enable proxy caching"
    )
    path: str = Field(
        default="/var/cache/nginx/cdn",
        description="Path to cache directory"
    )
    zone_name: str = Field(
        default="cdn_cache",
        description="Name of the cache zone"
    )
    zone_size: str = Field(
        default="100m",
        description="Size of cache keys zone"
    )
    max_size: str = Field(
        default="10g",
        description="Maximum size of cache on disk"
    )
    inactive: str = Field(
        default="7d",
        description="Time after which inactive cache is removed"
    )
    use_stale: List[str] = Field(
        default=["error", "timeout", "updating", "http_500", "http_502", "http_503", "http_504"],
        description="Conditions when stale cache can be used"
    )
    valid_codes: Dict[str, str] = Field(
        default={"200": "1d", "301": "1h", "302": "1h", "404": "1m"},
        description="Cache validity per response code"
    )


class NginxKeepaliveSettings(BaseModel):
    """Keepalive connection settings"""
    timeout: int = Field(
        default=65,
        ge=0, le=3600,
        description="Keepalive timeout for client connections (seconds)"
    )
    requests: int = Field(
        default=1000,
        ge=0, le=100000,
        description="Maximum requests per keepalive connection"
    )
    upstream_connections: int = Field(
        default=32,
        ge=0, le=1000,
        description="Number of keepalive connections to upstream"
    )


class NginxHttp2Settings(BaseModel):
    """HTTP/2 settings"""
    enabled: bool = Field(
        default=True,
        description="Enable HTTP/2 protocol"
    )
    max_concurrent_streams: int = Field(
        default=128,
        ge=1, le=1000,
        description="Maximum concurrent streams per connection"
    )
    max_field_size: str = Field(
        default="4k",
        description="Maximum size of a header field"
    )
    max_header_size: str = Field(
        default="16k",
        description="Maximum size of entire header list"
    )


class NginxSecuritySettings(BaseModel):
    """Security headers and settings"""
    server_tokens: bool = Field(
        default=False,
        description="Show nginx version in responses"
    )
    add_x_frame_options: bool = Field(
        default=True,
        description="Add X-Frame-Options: SAMEORIGIN header"
    )
    add_x_content_type_options: bool = Field(
        default=True,
        description="Add X-Content-Type-Options: nosniff header"
    )
    add_x_xss_protection: bool = Field(
        default=True,
        description="Add X-XSS-Protection header"
    )


class NginxRulesConfig(BaseModel):
    """Complete Nginx rules configuration"""
    client: NginxClientSettings = Field(default_factory=NginxClientSettings)
    websocket: NginxWebSocketSettings = Field(default_factory=NginxWebSocketSettings)
    proxy: NginxProxySettings = Field(default_factory=NginxProxySettings)
    gzip: NginxGzipSettings = Field(default_factory=NginxGzipSettings)
    ssl: NginxSSLSettings = Field(default_factory=NginxSSLSettings)
    rate_limit: NginxRateLimitSettings = Field(default_factory=NginxRateLimitSettings)
    cache: NginxCacheSettings = Field(default_factory=NginxCacheSettings)
    keepalive: NginxKeepaliveSettings = Field(default_factory=NginxKeepaliveSettings)
    http2: NginxHttp2Settings = Field(default_factory=NginxHttp2Settings)
    security: NginxSecuritySettings = Field(default_factory=NginxSecuritySettings)


class NginxRulesUpdate(BaseModel):
    """Partial update for Nginx rules"""
    client: Optional[NginxClientSettings] = None
    websocket: Optional[NginxWebSocketSettings] = None
    proxy: Optional[NginxProxySettings] = None
    gzip: Optional[NginxGzipSettings] = None
    ssl: Optional[NginxSSLSettings] = None
    rate_limit: Optional[NginxRateLimitSettings] = None
    cache: Optional[NginxCacheSettings] = None
    keepalive: Optional[NginxKeepaliveSettings] = None
    http2: Optional[NginxHttp2Settings] = None
    security: Optional[NginxSecuritySettings] = None


class NginxRulesResponse(BaseModel):
    """Response with current Nginx rules"""
    node_id: int
    node_name: str
    config: NginxRulesConfig
    last_applied: Optional[str] = None
    status: str = "unknown"


class NginxApplyResult(BaseModel):
    """Result of applying Nginx rules"""
    success: bool
    message: str
    config_test_output: Optional[str] = None
    reload_output: Optional[str] = None
